CORE-12937 schema_registry: core import mode support

pgellert opened the pull request:
Support enabling import mode in the Schema Registry.

In import mode, users can register schemas with the following differences from read write mode:
* users can specify the id and version of the schema
* re-inserting a schema definition with a different schema id is allowed
* no compatibility checks are performed

Fixes https://redpandadata.atlassian.net/browse/CORE-12937

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
schema_registry: allow enabling import mode

Don't reject IMPORT mode when parsing a request, instead, allow enabling
it.

The preconditions to enable import mode are:
- globally: the schema registry must be empty
- subject-level: the subject must be empty

Soft-deleted schema versions are also considered. The reference
implementation has a feature where it allows moving to import mode even
if there are soft-deleted schemas, and it hard-deletes these schemas
while moving into import mode. We can consider implementing this later
on (left as a TODO).

pgellert committed:
schema_registry: support import mode on register

Support IMPORT mode when handling `POST /subjects/{subject}/versions`
requests, that is, when registering schemas.

The expected behaviour is that in import mode:
* users can specify the id and version of the schema
* re-inserting a schema definition with a different schema id is allowed
* no compatibility checks are performed

pgellert committed:
rpk/registry: allow setting import mode

pgellert committed:
dt/schema_registry: test setting import mode

Test that the preconditions of enabling import mode are correctly
checked by the `PUT /mode` and `PUT /mode/{subject}` endpoints.

pgellert committed:
dt/schema_registry: support import mode on register

Test the behaviour of `POST /subjects/{subject}/versions` in import
mode.

pgellert committed:
schema_registry/errors: extract schema_id_already_exists

Extract for later reuse.

BenPope commented on src/go/rpk/pkg/cli/registry/mode/set.go:
```diff
@@ -23,7 +23,7 @@ import (
 )
 
-var supportedModes = []string{"READONLY", "READWRITE"}
+var supportedModes = []string{"READONLY", "READWRITE", "IMPORT"}
```
nit: Allowing this before it's supported seems to suggest the order of commits could be improved, it's not needed for tests?

BenPope commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -3964,22 +3964,22 @@ def test_mode(self):
         assert result_raw.status_code == requests.codes.ok
 
-        self.logger.debug("Set global mode to IMPORT")
+        self.logger.debug("Set global mode to FORWARD")
```
nit: This confused me a bit, since FORWARD is a supported config for compatibility. I'd be tempted to name it something like `UNSUPPORTED`.

BenPope commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -4130,6 +4130,144 @@ def test_mode_readonly(self):
         result_raw = self.sr_client.get_schemas_ids_id_versions(id=1)
         assert result_raw.status_code == 200
 
+    @cluster(num_nodes=3)
```
Is it worth adding a check that with global IMPORT, subject READ_WRITE behaves as intended (override?)?

pgellert replied:
I'm planning to have a follow up PR that removes our pseudo-import-mode behaviour in READ_WRITE mode. Let me defer adding testing for that to the follow up PR as well. I think it makes sense to test that behaviour both in global and subject-level READ_WRITE mode (with global IMPORT perhaps).

BenPope commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -4268,6 +4268,117 @@ def test_enabling_import_mode(self):
         self.assert_equal(result_raw.status_code, 200)
 
+    @cluster(num_nodes=3)
+    @matrix(subject_scope=[False, True])
```
I like this, could it apply to the previous test to reduce duplication?

pgellert replied:
`test_enabling_import_mode` can't be easily parametrized like this because it tests a sequence through different modes rather than two identical behaviours with subject-level and global mode

BenPope reviewed (commented):
Looks great.

I didn't see tests that our fake `IMPORT` in `READ_WRITE` mode has changed at all? I seem to recall that the reference implementation has some restrictions on specifying `id`/`version` during `READ_WRITE`.

Does this PR completely replace https://github.com/redpanda-data/redpanda/pull/23750?

I seem to recall an error code being produced from the reference implementation about version needing to be sequential? Perhaps about preventing version gaps? There was a specific error code for it, if I recall.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69746-->
<!--ccda1ac1-9622-4941-9ecf-e79e5c8a3d1f-->
<!--2736d086-cc00-4de4-85c8-dd81efd9fee0-->
<!--title-->
## Retry command for [Build#69746](https://buildkite.com/redpanda/redpanda/builds/69746)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/rpk_registry_test.py::RpkRegistryTest.test_registry_mode
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0411acff-2cb8-45ab-9952-d5456b85ea8a-->
<!--TEST RESULTS-->
<!--6c047aa8-6734-4204-a106-51a10eb48c5f-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69746">build#69746</a></summary>


| test_class                | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write           | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/69746#019850c4-b48e-41b2-8ada-add8af132313 | FLAKY         | 20/21    | upstream reliability is '94.77726574500768'. current run reliability is '95.23809523809523'. drift is -0.46083 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/69746#019850c4-b48d-4650-be2e-2a917ce54997 | FLAKY         | 19/21    | upstream reliability is '99.38837920489296'. current run reliability is '90.47619047619048'. drift is 8.91219 and the allowed drift is set to 50. The test should PASS  |
| RpkRegistryTest           | test_registry_mode   | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/69746#019850c4-b48b-4bc6-a61a-41fec1a0eecf | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| RpkRegistryTest           | test_registry_mode   | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/69746#019850c4-bf8f-4fb2-853e-38c1ad8ddcaa | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69925">build#69925</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/69925#01985a69-c4de-4ca9-9bf8-f48d9115d846 | FLAKY         | 20/21    | upstream reliability is '99.24242424242425'. current run reliability is '95.23809523809523'. drift is 4.00433 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

r-vasquez commented:
rpk bits lgtm, we need to change this though: https://github.com/redpanda-data/redpanda/blob/dev/tests/rptest/tests/rpk_registry_test.py#L654-L657 😅 

We wanted to test this because rpk was the one who made the check because the underlying library already accepted the `IMPORT` mode.

pgellert committed:
schema_registry: return 42205 when schema id exists

The error "Overwrite new schema with id 3 is not permitted." should use
error code 42205 to match the reference implementation.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/66fedfbf5a0116a49448f377353c5bb739af33f3..43c01e7053ee9a506410ff395f3ef034446d9aae):
* address feedback
* fix rpk test failure
* add https://github.com/redpanda-data/redpanda/pull/27010/commits/6bd36ae02af27700be4f09eeb53aa8f5be9cbc39 -- change error status from 42207 to 42205 when the error is "Overwrite new schema with id {} is not permitted."

pgellert commented:
> I didn't see tests that our fake `IMPORT` in `READ_WRITE` mode has changed at all? I seem to recall that the reference implementation has some restrictions on specifying `id`/`version` during `READ_WRITE`.

I have a separate ticket for changing the `READ_WRITE` mode to align with the reference implementation, so that will be the next PR.

> Does this PR completely replace #23750?

Yes. Thanks for linking to it, now I noticed that we should change the error code from 42207 to 42205 when returning an error on overwriting a schema, so I've included that in this PR as well (https://github.com/redpanda-data/redpanda/pull/27010/commits/6bd36ae02af27700be4f09eeb53aa8f5be9cbc39).

> I seem to recall an error code being produced from the reference implementation about version needing to be sequential? Perhaps about preventing version gaps? There was a specific error code for it, if I recall.

In IMPORT mode the version can be arbitrary, the requirement on the version is only a requirement in READ_WRITE mode. The error you're thinking of is `{"error_code":42201,"message":"Version is not one more than previous version"}`. I'll implement that in the next PR.

BenPope commented on src/v/pandaproxy/schema_registry/seq_writer.cc:
```diff
@@ -391,6 +391,35 @@ ss::future<std::optional<bool>> seq_writer::do_write_mode(
+            }
+        }
+
+        // TODO: relax the above restrictions to
+        // 1. Allow soft-deleted schemas to exist, but
+        // 2. Hard delete them before moving to import mode
```
I guess this would be conditional on `force`?

pgellert replied:
The code would come here where the todo is, so it would be skipped if `force=true`.

BenPope commented on src/v/pandaproxy/schema_registry/seq_writer.cc:
```diff
@@ -224,7 +224,13 @@ void seq_writer::advance_offset_inner(model::offset offset) {
+    co_await check_mutable(sub);
+
+    const auto mode = co_await _store.get_mode(sub, default_to_global::yes);
+    if (schema.id < 0 && mode != mode::read_write) {
```
Is there any significance of the `id` being negative but not `invalid_schema_id`?

pgellert replied:
No, any negative number is treated the same as `invalid_schema_id` / not specified.
